src/index.ts

```diff
@@ -13,49 +13,100 @@
+          err ? resolve(false) : resolve(true);
+        });
+  });
+}
+
+// Returns value of metadata field.
+// Throws error if there is a problem accessing metadata API.
+function getField(field: string): Promise<string> {
+  return new Promise<string>((resolve, reject) => {
+    request.get(
+        {url: metadataAPI + field, headers: {'Metadata-Flavor': 'Google'}},
+        function(err: Error, response: request.RequestResponse) {
+          err ? reject(err) : resolve(response.body);
+        });
+  });
+}
 
 // initConfig sets unset values in the configuration to the value retrieved from
-// environment variables, metadata, or the default value specified in
+// environment variables, metadata, or the default values specified in
 // defaultConfig.
-// Returns rejected promise if value that must be set cannot be initialized.
-async function initConfig(config: Config): Promise<ProfilerConfig> {
+// Throws error if value that must be set cannot be initialized.
+// Exported for testing purposes.
+export async function initConfig(config: Config): Promise<ProfilerConfig> {
   config = common.util.normalizeArguments(null, config);
 
-  const envConfig = {
-    logLevel: process.env.GCLOUD_PROFILER_LOGLEVEL,
+  const envConfig: any = {
     projectId: process.env.GCLOUD_PROJECT,
     serviceContext: {
       service: process.env.GAE_SERVICE,
       version: process.env.GAE_VERSION,
     }
   };
 
+  if (process.env.GCLOUD_PROFILER_LOGLEVEL !== undefined) {
+    let envLogLevel = parseInt(process.env.GCLOUD_PROFILER_LOGLEVEL || '', 10);
+    if (envLogLevel !== NaN) {
+      envConfig.logLevel = envLogLevel;
+    }
+  }
+
   let envSetConfig: Config = {};
   if (process.env.hasOwnProperty('GCLOUD_PROFILER_CONFIG')) {
     envSetConfig =
         require(path.resolve(process.env.GCLOUD_PROFILER_CONFIG)) as Config;
   }
 
-  let normalizedConfig = extend(true, {}, defaultConfig, envConfig, config);
+  let normalizedConfig =
+      extend(true, {}, defaultConfig, envSetConfig, envConfig, config);
 
-  if (normalizedConfig.serviceContext.service === undefined) {
-    throw new Error('service name must be specified in the configuration');
+  // fetch instance and zone from metadata if not specified and on GCE
+  if (await onGCE()) {
+    if (!normalizedConfig.instance) {
+      normalizedConfig.instance = await getField('instance/name').catch((e) => {
+        throw Error('failed to get instance from Compute Engine: ' + e);
+      });
+    }
+    if (!normalizedConfig.zone) {
+      let zone = await getField('instance/zone').catch((e) => {
+        throw Error('failed to get zone from Compute Engine: ' + e);
+      });
+      normalizedConfig.zone = zone.substring(zone.lastIndexOf('/') + 1);
+    }
   }
 
-  // TODO: fetch instance and zone from metadata. This will require function to
-  // be asynchrous.
-  if (normalizedConfig.instance === undefined) {
-    normalizedConfig.instance = '';
+  if (normalizedConfig.projectId === undefined) {
```

Comment:
Requiring the projectId from the user is too aggressive compared to the other agents. The authentication flow will automatically figure out the project id (or give an error) if one is not provided.